### PR TITLE
Replace recipients list with bloom filter

### PR DESCRIPTION
**Pull request by jrandallclark**
**Description**

Replaces recipients list with bloom filter.
Consumer fetches individual content keys after fetching the file instead of producer putting them in file metadata.

### CODE REVIEW

File: app/src/main/java/memphis/myapplication/UI/LoginFragment.java [hunk 1]
```diff
@@ -162,6 +162,7 @@ private int loginAttempt(String domain, String username, String password) {
         protected Void doInBackground(Void... voids) {
             // save username and go to mainpage
             SharedPrefsManager.getInstance(getActivity()).setCredentials(username, password, domain);
+            Globals.setUsername(username);
             return null;
```
Comment (agawande) [inline app/src/main/java/memphis/myapplication/UI/LoginFragment.java:166]: Is return null needed here?

File: app/src/main/java/memphis/myapplication/data/BackgroundJobs.java [hunk 2]
```diff
@@ -238,6 +222,7 @@ public void setup_security(Context applicationContext) {
         Globals.setFace(face);
         Globals.setMemoryCache(new MemoryCache(face, applicationContext));
         sharedPrefsManager.generateKey();
+        Globals.setUsername(sharedPrefsManager.getUsername());
```
Comment (agawande) [inline app/src/main/java/memphis/myapplication/data/BackgroundJobs.java]: Why is username set here and in LoginFragment?

File: app/src/main/java/memphis/myapplication/data/Common.java [hunk 3]
```diff
@@ -327,38 +325,38 @@ public static void encryptAndPublish(Intent resultData, Context context, boolean
...
+
+                    metadata.setFeed(false);
                     Timber.d("For friends");
-                    for (String friend : recipients) {
-                        Blob friendKey = databaseViewModel.getFriend(friend).getCert().getPublicKey();
-                        byte[] encryptedKey = RsaAlgorithm.encrypt
-                                (friendKey, new Blob(secretKey.getEncoded()), new EncryptParams(EncryptAlgorithmType.RsaOaep)).getImmutableArray();
-                        syncData.addFriendKey(friend, encryptedKey);
-                    }
+                    BloomFilter bloomFilter = new BloomFilter(recipients.size(), 0.001);
```
Comment (agawande) [inline app/src/main/java/memphis/myapplication/data/Common.java]: Probably want to have 0.001 as a constant somewhere or maybe have 0.001 as default and another constructor which only takes the size.

File: app/src/main/java/memphis/myapplication/data/Common.java [hunk 4]
```diff
@@ -327,38 +325,38 @@ public static void encryptAndPublish(Intent resultData, Context context, boolean
...
+
+                    Name bloomName = new Name(bloomFilter.appendToName(new Name()));
+                    Timber.d("Size: " + bloomName.get(0).toNumber() + " Prob: " + bloomName.get(1).toNumber());
+                    metadata.setBloomFilter(bloomName);
+
                 }
-                filename = filename + "/" + Common.getKeyDigest(secretKey);
+                String keyDigest = Common.getKeyDigest(secretKey);
+                PublishedContent publishedContent = databaseViewModel.checkIfShared(keyDigest);
+                filename = filename + "/" + keyDigest;
```
Comment (agawande) [inline app/src/main/java/memphis/myapplication/data/Common.java:357]: Is there some inbuilt class that handles creating file names?

File: app/src/main/java/memphis/myapplication/data/RealmRepository.java [hunk 5]
```diff
@@ -214,6 +214,7 @@ public SecretKey getSymKey(String friendName) {
             User user = userRealmToUser(realm.where(UserRealm.class).equalTo("username", friendName).findFirst());
             byte[] encryptedKey = user.getSymKey();
             realm.commitTransaction();
+            if (encryptedKey == null) return null;
```
Comment (agawande) [inline app/src/main/java/memphis/myapplication/data/RealmRepository.java:217]: Usually discourage these types of formatting in ndn cxx code style. Not sure about Java/jndn.
https://named-data.net/doc/ndn-cxx/current/code-style.html

File: app/src/main/java/memphis/myapplication/data/tasks/FetchingTask.java [hunk 6]
```diff
@@ -309,30 +307,34 @@ public void succeed() {
...
     }
 
-    public void fetchKey() {
-       requestSymKey(RealmRepository.getInstanceForNonUI().getFriend(m_user).getNamespace(), m_fileKeyDigest, SharedPrefsManager.getInstance(m_currContext).getUsername());
+    public void fetchKey(String keyDigest) {
+       requestSymKey(RealmRepository.getInstanceForNonUI().getFriend(m_user).getNamespace(), keyDigest, SharedPrefsManager.getInstance(m_currContext).getUsername());
     }
 
-    public void requestSymKey(final String friendNameSpace, String keyName, String username) {
+    public void requestSymKey(final String friendNameSpace, final String keyDigest, String username) {
```
Comment (agawande) [inline app/src/main/java/memphis/myapplication/data/tasks/FetchingTask.java]: Why is username not final?

File: app/src/main/java/memphis/myapplication/psync/ConsumerManager.java [hunk 7]
```diff
@@ -90,35 +78,26 @@ public void onData(Interest interest, Data data) {
...
-                    RealmRepository feedLoopRealmRepository = RealmRepository.getInstanceForNonUI();
-                    Timber.d(interest.getName().toUri());
-                    Timber.d("Friend name: " + friendName);
-                    Timber.d(feedLoopRealmRepository.getSymKey(friendName).toString());
-                    new FetchingTask(context, toastData).execute(new FetchingTaskParams(new Interest(new Name(filename)), feedLoopRealmRepository.getSymKey(friendName), true));
+                    new FetchingTask(context, toastData).execute(new FetchingTaskParams(new Interest(new Name(filename)), true));
 
                 } else {
-                    if (syncData.forMe(SharedPrefsManager.getInstance(context).getUsername())) {
+                    BloomFilter bloomFilter = new BloomFilter(metadata.getBloomFilter().get(0).toNumber(), 0.001, metadata.getBloomFilter().get(-1));
```
Comment (agawande) [inline app/src/main/java/memphis/myapplication/psync/ConsumerManager.java]: Same as 0.001 comment above - probably best to move it as default value inside BloomFilter and provide various different ctors.

File: app/src/main/java/memphis/myapplication/utilities/BloomFilter.java [hunk 8]
```diff
@@ -0,0 +1,387 @@
...
+    static BloomParameters
+    getParameters(long projected_element_count,
+                  double false_positive_probability)
+    {
+        BloomParameters opt = new BloomParameters();
+        opt.false_positive_probability = false_positive_probability;
+        opt.projected_element_count = projected_element_count;
+
+        if (opt.not()) {
+            System.out.println("Bloom parameters are not correct!");
```
Comment (agawande) [inline app/src/main/java/memphis/myapplication/utilities/BloomFilter.java]: Shouldn't this be replaced by Timber?

File: app/src/main/java/memphis/myapplication/utilities/BloomFilter.java [hunk 9]
```diff
@@ -0,0 +1,387 @@
...
+    }
+
+    public static long toUnsignedLong(int x) {
+        return x & 0x00000000ffffffffL;
+    }
+
+    public static int toUnsignedInt(byte x) { return x & 0xFF; }
+
+}
+class BloomParameters
```
Comment (agawande) [inline app/src/main/java/memphis/myapplication/utilities/BloomFilter.java:335]: All of this class is c style formatted. Should we format more like Java?

File: app/src/test/java/memphis/myapplication/TLVEncodeDecodeTest.java [hunk 10]
```diff
@@ -1,35 +0,0 @@
...
-
-import javax.crypto.SecretKey;
-
-import memphis.myapplication.UI.MainFragment;
-import memphis.myapplication.psync.ConsumerManager;
-import memphis.myapplication.data.tasks.FetchingTaskParams;
-
-import static org.junit.Assert.assertThat;
-
-public class TLVEncodeDecodeTest {
```
Comment (agawande) [inline app/src/test/java/memphis/myapplication/TLVEncodeDecodeTest.java:22]: Why is this deleted?

**Review by agawande — CHANGES_REQUESTED** [review]
See single line comments.

File: app/src/test/java/memphis/myapplication/TLVEncodeDecodeTest.java (hunk 10 above)
Comment (jrandallclark) [reply app/src/test/java/memphis/myapplication/TLVEncodeDecodeTest.java:22]: We switched to JSON for the file metadata over the summer and aren't using TLV anymore.

File: app/src/main/java/memphis/myapplication/UI/LoginFragment.java (hunk 1 above)
Comment (jrandallclark) [reply app/src/main/java/memphis/myapplication/UI/LoginFragment.java:166]: This is from one of the GSoC applicants, but from I believe so. doInBackground has to return something. The method's return type is Void, so it may be possible to return Void. Either way, it's never touched again.

File: app/src/main/java/memphis/myapplication/data/BackgroundJobs.java (hunk 2 above)
Comment (jrandallclark) [reply app/src/main/java/memphis/myapplication/data/BackgroundJobs.java]: Leftover from testing. Will remove.

File: app/src/main/java/memphis/myapplication/data/Common.java (hunk 4 above)
Comment (jrandallclark) [reply app/src/main/java/memphis/myapplication/data/Common.java:357]: No. But there probably should be. A filename is <user-namespace>/file/<path>/<key-digest>.

File: app/src/main/java/memphis/myapplication/utilities/BloomFilter.java (hunk 9 above)
Comment (jrandallclark) [reply app/src/main/java/memphis/myapplication/utilities/BloomFilter.java:335]: I can. What are the major differences other than, say, like breaks in the method headers?
